reference/reference.go

```diff
@@ -0,0 +1,280 @@
+// Package reference provides a general type to represent any way of referencing images within the registry.
+// Its main purpose is to abstract tags and digests (content-addressable hash).
+//
+// Grammar
+//
+//     reference                       := repository [ ":" tag ] [ "@" digest ]
+//     repository                      := component [ [ port ] ( '/' component )* ]
+//     component                       := alpha-numeric [ separator alpha-numeric ]*
+//     alpha-numeric                   := /[a-z0-9]+/
+//     separator                       := /[._-]/
+//     port                            := /:[0-9]+/
+//     tag                             := /[\w][\w.-]{0,127}/
+//
+//     // from the digest package
+//     digest                          := digest-algorithm ":" digest-hex
+//     digest-algorithm                := digest-algorithm-component [ digest-algorithm-separator digest-algorithm-component ]
+//     digest-algorithm-separator      := /[+.-_]/
+//     digest-algorithm-component      := /[A-Za-z]/ /[A-Za-z0-9]*/
+//     digest-hex                      := /[A-Za-z0-9_-]+/ ; supports hex bytes or url safe base64
+package reference
+
+import (
+	"errors"
+	"regexp"
+
+	"github.com/docker/distribution/digest"
+)
+
+var (
+	tagRegexp    = regexp.MustCompile(`[\w][\w.-]{0,127}`)
+	alphaNumeric = regexp.MustCompile(`[a-z0-9]+`)
+	portRegexp   = regexp.MustCompile(`:[0-9]+`)
+)
+
+// ErrReferenceInvalidFormat represents an error while trying to parse a string as a reference.
+var ErrReferenceInvalidFormat = errors.New("invalid reference format")
+
+// Reference abstracts types that reference images in a certain way.
+type Reference interface {
+	// Returns the repository part of a reference
+	Repository() string
+	// Returns the entire reference, including the repository
+	String() string
+}
+
+type parser struct {
+	err error
+}
+
+// parseComponent parses a component of the repository.
+//
+// 	component                       := alpha-numeric [separator alpha-numeric]*
+// 	separator                       := /[._-]/
+func (p *parser) parseComponent(s string) (component, tail string) {
```

Comment:
@stevvooe i believe it's always right, because if a deep parser is failing, it cannot know if its caller should take that error as a hard error, or if it can branch to an alternative codepath. But feel free to prove me wrong